app/assets/javascripts/TortoiseJS/agent/view.coffee

```diff
@@ -369,3 +283,251 @@ class PatchDrawer
     if world.patcheswithlabels
       @labelPatches(patches)
 
+
+class Line
+  constructor: (@x1, @y1, @x2, @y2) ->
+
+  midpoint: ->
+    midpointX = (@x1 + @x2) / 2
+    midpointY = (@y1 + @y2) / 2
+    [midpointX, midpointY]
+
+class LinkDrawer
+  constructor: (@view, @shapes) ->
+    directionIndicators = {}
+    for name, shape of @shapes
+      directionIndicators[name] = shape['direction-indicator']
+    @linkShapeDrawer = new ShapeDrawer(directionIndicators)
+
+  drawCurvedLine: (x1, y1, x2, y2, cx, cy) =>
+    @view.ctx.moveTo(x1, y1)
+    @view.ctx.quadraticCurveTo(cx, cy, x2, y2)
+
+  drawLine: (x1, y1, x2, y2) =>
+    @view.ctx.moveTo(x1, y1)
+    @view.ctx.lineTo(x2, y2)
+
+  shouldWrapInDim: (canWrap, dimensionSize, cor1, cor2) ->
+    distance = Math.abs(cor1 - cor2)
+    canWrap and distance > dimensionSize / 2
+
+  adjustLinkEnds: (link, end1Cor, end2Cor, trigShift, minForExtending, maxForExtending) ->
+    diff = Math.abs(link.thickness * trigShift) / 2
+    if minForExtending < link.heading < maxForExtending
+      [end1Cor + diff, end2Cor - diff]
+    else
+      [end1Cor - diff, end2Cor + diff]
```

Comment:
This logic doesn't take into account curviness. In the model I posted, `ask turtles [ st ]`. You'll see the turtles are very off-center. This is because the above logic pushes the link ends to the sides by `thickness`.

Turn to the darkside. Just use circles.